### PR TITLE
Update Fine_Tuning_DenseNet.ipynb

### PR DESCRIPTION
**Pull request by allaedding**
after feature-Extraction you need to work on the trained model "3 epochs" not on the first model, If I Am wrong  please clarify this part for me 
you did train your model for 3 epoch to help initialize the weights then you freeze the weights on the first one then you passed the trained model "3 epochs" to the final 30 epochs training .

### CODE REVIEW

**Comment by hoang-ho** [comment]
Hi! Yes, That's exactly what I did.

**Comment by hoang-ho** [comment]
Sorry for not correctly addressing your question previously. Before that I defined:

```
last_layer = pre_trained_model.get_layer('relu')
print('last layer output shape:', last_layer.output_shape)
last_output = last_layer.output

# Flatten the output layer to 1 dimension
x = layers.GlobalMaxPooling2D()(last_output)
# Add a fully connected layer with 512 hidden units and ReLU activation
x = layers.Dense(512, activation='relu')(x)
# Add a dropout rate of 0.7
x = layers.Dropout(0.5)(x)
# Add a final sigmoid layer for classification
x = layers.Dense(7, activation='softmax')(x)

model = Model(pre_trained_model.input, x)
```

What I want to freeze is the pre_trained_model's layers, hence:

```
for layer in pre_trained_model.layers[481:]:
    layer.trainable = True
```

This is not the same as:

```
for layer in model.layers[481:]:
    layer.trainable = True
```

because, model includes all layers of pre_trained_model plus 4 additional top layers, which I defined on my own for this specific classification task. 
Remember that, I didn't include the top layers when importing pre_trained_model:

```
pre_trained_model = DenseNet201(input_shape=(192, 256, 3), include_top=False, weights="imagenet")
```
I hope this is helpful

**Comment by allaedding** [comment]
thank you for the reply, I understand the difference between the two models, But I didn't understand this, see in the picture the numbers :
"model" = "pre_trained_model" plus 4
in "1" you trained the "model" for feature extraction, ok 
in "2 " you have freezed the layers after layer 481, for the "pre_trained_model" and that's not the one you trained for feature extraction.
in "3" and "4" you compiled an train the "model" for 30 epochs but not the one with freezed layers.

I don't understand this how you have freezed the layers to "pre_trained_model" and that affect the layers in the "model".



### **see the code for your vgg16 fine-tuning you freeze the layers for the same "model" you trained for feature extraction.**




Capture

**Comment by hoang-ho** [comment]
I defined:
```
last_layer = pre_trained_model.get_layer('relu')
print('last layer output shape:', last_layer.output_shape)
last_output = last_layer.output

# Flatten the output layer to 1 dimension
x = layers.GlobalMaxPooling2D()(last_output)
# Add a fully connected layer with 512 hidden units and ReLU activation
x = layers.Dense(512, activation='relu')(x)
# Add a dropout rate of 0.7
x = layers.Dropout(0.5)(x)
# Add a final sigmoid layer for classification
x = layers.Dense(7, activation='softmax')(x)

model = Model(pre_trained_model.input, x) # this is the definition of model
```

Basically, model already includes all layers of the pre_trained_model plus a few top layers. When I trained it for 3 epochs first, weights of newly defined top layers are changed but not the pre_trained_model's layers (except for batch norm layers, refer to technical issue). Then I just unfroze the very last dense block in the pre_trained_model, which effectively unfroze the corresponding dense block in model. You can think of pre_trained_model.layers (for example the one in VGG16) as a list containing the following objects: input_1, block1_conv1, block1_conv2, block1_pool, ..., block5_conv1, block5_conv2, block5_conv3, block5_pool. model.layers is a list containing those same objects plus a few more. 

I'm sorry for the inconsistency in the VGG16 script, the following

```
for layer in pre_trained_model.layers[15:]:
    layer.trainable = True
```

would achieve the same thing.

**Comment by allaedding** [comment]
Sorry I understand my mistake now, thank you for being so patient with me and for your kind reply.
